web/src/echo.ts

```diff
@@ -21,17 +24,84 @@ import {current_user} from "./state_data";
 import * as stream_data from "./stream_data";
 import * as stream_list from "./stream_list";
 import * as stream_topic_history from "./stream_topic_history";
+import type {TopicLink} from "./types";
 import * as util from "./util";
 
 // Docs: https://zulip.readthedocs.io/en/latest/subsystems/sending-messages.html
 
-const waiting_for_id = new Map();
-let waiting_for_ack = new Map();
+type ServerMessage = RawMessage & {local_id?: string};
+
+const send_message_api_response_schema = z.object({
+    id: z.number(),
+    automatic_new_visibility_policy: z.number().optional(),
+});
+
+type MessageRequestObject = {
+    sender_id: number;
+    queue_id: null | string;
+    topic: string;
+    content: string;
+    to: string;
+    draft_id: string | undefined;
+};
+
+type PrivateMessageObject = {
+    type: "private";
+    reply_to: string;
+    private_message_recipient: string;
+    to_user_ids: string | undefined;
+};
+
+type StreamMessageObject = {
+    type: "stream";
+    stream_id: number;
+};
+
+type MessageRequest = MessageRequestObject & (PrivateMessageObject | StreamMessageObject);
+
+type LocalEchoRequest = Partial<{
+    raw_content: string | undefined;
+    content: string;
+    orig_content: string;
+    orig_raw_content: string | undefined;
+    new_topic: string;
+    new_stream_id: number;
+    starred: boolean;
+    historical: boolean;
+    collapsed: boolean;
+    alerted: boolean;
+    mentioned: boolean;
+    mentioned_me_directly: boolean;
+}>;
```

Comment:
Renamed this to `LocalEditRequest`, since it's for locally echoing the EDIT of an existing message.